### PR TITLE
Align momentum optimizer with conventional parameterization and add Nesterov

### PR DESCRIPTION
**Pull request by jekbradbury**
The implementation of SGD with momentum in optimizers.py was different from the typical parameterization. Thanks to @sschoenholz for noticing!

For background, there are two common parameterizations of ordinary momentum:
- Keras (and thus TF 2.X) use `v = mu * v - eta * g; x = x + v` (which expands to e.g. `x3 = x0 - (eta0*mu^2 + eta0*mu + eta0)*g0 - (eta1*mu + eta1)*g1 - eta2*g2`)
- PyTorch and Sonnet (and TF 1.X) use `v = mu * v + g; x = x - eta * v` (which expands to `x3 = x0 - (eta2*mu^2 + eta1*mu + eta0)*g0 - (eta2*mu + eta1)*g1 - eta2*g2`)

Other than their representation of internal state `v`, those expansions show that these two formulas only differ in how they're affected by changing learning rates. There are two analogous implementations for Nesterov momentum, again equivalent other than in the case of learning rate changes. The original papers describing momentum and Nesterov momentum assume constant learning rate, so they're ambiguous between the formulations.

The version of momentum in optimizers.py was different (related by a reparameterization of the learning rate hyperparameter). With this change, it exactly matches the formulas used by PyTorch (when `damping=0`) and Sonnet; we also add a Nesterov implementation.

### CODE REVIEW

**Comment by mattjj** [comment]
See also HIPS/autograd#289 for a discussion.

I wouldn't say the implementation currently optimizers.py is _wrong_ just because it uses a different parameterization. In that Autograd issue I showed how to map a fixed learning rate from the TF/Keras convention to the Autograd/JAX convention. Here's a version of the same for JAX:

```python
from __future__ import print_function
import numpy as np

gs = np.random.randn(10)

def tf_momentum(x0, v0, eta, mu):
    x = x0
    v = v0
    for g in gs:
        v = mu * v - eta * g
        x = x + v
        print(x)
    print()

def jax_momentum(x0, v0, eta, mu):
    x = x0
    v = v0
    for g in gs:
        v = mu * v - (1.0 - mu) * g
        x = x + eta * v
        print(x)
    print()

tf_momentum(0., 0., 0.5, 0.9)
jax_momentum(0., 0., 0.5 / (1. - 0.9), 0.9)
```

```
0.02812472364200032
0.7963849340762899
1.4299981786941829
1.3300049027532435
0.8492142695935796
-0.08592686603497257
-0.830986524831802
-1.0853028959102888
-1.167070274778137
-1.2185173417800437

0.028124723642000315
0.79638493407629
1.4299981786941833
1.3300049027532443
0.8492142695935807
-0.08592686603497135
-0.8309865248318008
-1.0853028959102877
-1.1670702747781359
-1.2185173417800428
```

But I agree with your point (I think) that it's good to follow conventions. The downside now is this will change the behavior of users' existing code. One option is to delay this change until we make stax2 / optimizers2. But this might be minor enough that we can just make the change now, so long as we warn people in JAXers chat. What do you think?

(Btw the cocalc link didn't load for me.)

File: jax/experimental/optimizers.py [hunk 1]
```diff
@@ -200,8 +200,8 @@ def get_params(x):
   return init, update, get_params
 
 @optimizer
-def momentum(step_size, mass):
-  """Construct optimizer triple for SGD with Nesterov momentum.
+def momentum(step_size, mass, nesterov=False):
```
Comment (mattjj) [inline jax/experimental/optimizers.py]: I'm a little hesitant of switching behavior based on a flag, and doubly hesitant about that flag having a default value. How about we just have two separate optimizers? Each is only a few lines, and that will make the comparison even easier (and avoid the `d` intermediate). WDYT?

**Comment by mattjj** [comment]
@froystig is there a canonical reference we can link to in our comments / docstrings?

**Comment by jekbradbury** [comment]
I completely missed the Autograd connection here (in retrospect I should have looked there first!) and I also misread a sign while trying to relate the JAX and TF formulations (so it looked like they weren’t reparameterizations, even though @sschoenholtz had pointed out the relationship to me already). While I can see conceptual arguments for either basic formulation (e.g., is the step size controlling the instantaneous effect of the gradient or multiplying the overall velocity?) it looks like the ship has probably sailed in favor of the learning rate parameterization used in TF and PyTorch. On the other hand it’s hard to choose between the TF and PyTorch parameterizations of the velocity, since the difference is subtle and people’s expectations are split...

**Comment by mattjj** [comment]
Maybe we can check `sys.path` to see if the user has PyTorch or TensorFlow installed, then switch our convention based on that. :D
